File: Python/OneLang/One/Transforms/FillMutabilityInfo.py
from OneLangStdLib import *
import OneLang.One.AstTransformer as astTrans
import OneLang.One.Ast.Types as types
import OneLang.One.Ast.Expressions as exprs
import OneLang.One.Ast.References as refs
import OneLang.One.Ast.Statements as stats

class FillMutabilityInfo(astTrans.AstTransformer):
    def __init__(self):
        super().__init__("FillMutabilityInfo")
    
    def get_var(self, var_ref):
        v = var_ref.get_variable()
        if v.mutability == None:
            v.mutability = types.MutabilityInfo(True, False, False)
        return v
    
    def visit_variable_reference(self, var_ref):
        self.get_var(var_ref).mutability.unused = False
        return var_ref
    
    def visit_variable_declaration(self, stmt):
        super().visit_variable_declaration(stmt)
        if stmt.attributes != None and stmt.attributes.get("mutated") == "true":
            stmt.mutability.mutated = True
        return stmt
    
    def visit_expression(self, expr):
        expr = super().visit_expression(expr)
        
        if isinstance(expr, exprs.BinaryExpression) and isinstance(expr.left, refs.VariableReference) and expr.operator == "=":
            self.get_var(expr.left).mutability.reassigned = True
        elif isinstance(expr, exprs.InstanceMethodCallExpression) and isinstance(expr.object, refs.VariableReference) and "mutates" in expr.method.attributes:
            self.get_var(expr.object).mutability.mutated = True
        return expr
    
    def visit_variable(self, variable):
        if variable.mutability == None:
            variable.mutability = types.MutabilityInfo(True, False, False)
        return variable

File: PHP/src/OneLang/One/Transforms/FillMutabilityInfo.php
<?php

namespace OneLang\One\Transforms\FillMutabilityInfo;

use OneLang\One\AstTransformer\AstTransformer;
use OneLang\One\Ast\Types\IVariable;
use OneLang\One\Ast\Types\MutabilityInfo;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\BinaryExpression;
use OneLang\One\Ast\Expressions\InstanceMethodCallExpression;
use OneLang\One\Ast\References\VariableReference;
use OneLang\One\Ast\Statements\VariableDeclaration;

class FillMutabilityInfo extends AstTransformer {
    function __construct() {
        parent::__construct("FillMutabilityInfo");
        
    }
    
    protected function getVar($varRef) {
        $v = $varRef->getVariable();
        if ($v->mutability === null)
            $v->mutability = new MutabilityInfo(true, false, false);
        return $v;
    }
    
    protected function visitVariableReference($varRef) {
        $this->getVar($varRef)->mutability->unused = false;
        return $varRef;
    }
    
    protected function visitVariableDeclaration($stmt) {
        parent::visitVariableDeclaration($stmt);
        if ($stmt->attributes !== null && @$stmt->attributes["mutated"] ?? null === "true")
            $stmt->mutability->mutated = true;
        return $stmt;
    }
    
    protected function visitExpression($expr) {
        $expr = parent::visitExpression($expr);
        
        if ($expr instanceof BinaryExpression && $expr->left instanceof VariableReference && $expr->operator === "=")
            $this->getVar($expr->left)->mutability->reassigned = true;
        else if ($expr instanceof InstanceMethodCallExpression && $expr->object instanceof VariableReference && array_key_exists("mutates", $expr->method->attributes))
            $this->getVar($expr->object)->mutability->mutated = true;
        return $expr;
    }
    
    protected function visitVariable($variable) {
        if ($variable->mutability === null)
            $variable->mutability = new MutabilityInfo(true, false, false);
        return $variable;
    }
}
